src/components/FileNameHeader.tsx

```diff
@@ -0,0 +1,136 @@
+import { Pencil1Icon } from '@radix-ui/react-icons'
+import {
+  Badge,
+  Button,
+  Dialog,
+  Flex,
+  Heading,
+  IconButton,
+  TextField,
+  Tooltip,
+} from '@radix-ui/themes'
+import { useEffect, useRef, useState } from 'react'
+import { useForm } from 'react-hook-form'
+
+import { useOverflowCheck } from '@/hooks/useOverflowCheck'
+import { useRenameFile } from '@/hooks/useRenameFile'
+import { StudioFile } from '@/types'
+import { getFileExtension } from '@/utils/file'
+
+import { FieldGroup } from './Form'
+
+interface FileNameHeaderProps {
+  file: StudioFile
+  isDirty?: boolean
+  showExt?: boolean
+}
+
+export function FileNameHeader({
+  file,
+  isDirty = false,
+  showExt = false,
+}: FileNameHeaderProps) {
+  const [isOpen, setIsOpen] = useState(false)
+  const subTitleRef = useRef<HTMLHeadingElement>(null)
+  const hasEllipsis = useOverflowCheck(subTitleRef)
+  const fileExtension = getFileExtension(file.fileName)
+
+  return (
+    <>
+      <Tooltip content={file.displayName} hidden={!hasEllipsis}>
+        <Heading
+          size="2"
+          weight="medium"
+          color="gray"
+          truncate
+          ref={subTitleRef}
+        >
+          {file.displayName}
+        </Heading>
+      </Tooltip>
+
+      {isDirty && (
+        <Tooltip content="Unsaved changes">
+          <span aria-label="Unsaved changes">*</span>
+        </Tooltip>
+      )}
+
+      <Tooltip content="Rename file">
+        <IconButton
+          variant="ghost"
+          color="gray"
+          aria-label="Rename file"
+          onClick={() => setIsOpen(true)}
+        >
+          <Pencil1Icon />
+        </IconButton>
+      </Tooltip>
+
+      <RenameFileDialog file={file} open={isOpen} onOpenChange={setIsOpen} />
+
+      {showExt && !!fileExtension && (
+        <Badge color="gray" size="1">
+          {fileExtension.toUpperCase()}
+        </Badge>
+      )}
+    </>
+  )
+}
+
+interface RenameFileDialogProps {
+  file: StudioFile
+  open: boolean
+  onOpenChange: (open: boolean) => void
+}
+
+function RenameFileDialog({ file, open, onOpenChange }: RenameFileDialogProps) {
```

Comment:
I wanted to have some kind of an inline text input, but didn't have time to implement it the right way 🫤